### DISCUSSION TITLE
Use DI to dispose of database connections

### DISCUSSION THREAD

**Discussion by leo-amelyan** [JasperFx/marten#2084] [Q&A]
I got an issue where connections were not being disposed of. One way to fix these is by manually disposing of connections however, this is error-prone and I was wondering if there is a better approach.

[In this post](https://github.com/JasperFx/marten/issues/1215#issuecomment-510947218), it is suggested to use DI configuration:

>I'd personally recommend to not do manual disposal of the DocumentSession but let the Dependency Injection container let manage it properly - so eg. use request scope registration for DocumentSession and Singleton for Document Store.

How would I configure this? 

I am using Lamar for IoC. Referencing [this Marten documentation page](https://martendb.io/configuration/ioc.html), I created a MartenServices which configures the IDocumentStore as singleton and IDocumentSession as scoped. However, when I test the application, the database connections are still not disposed of.

**Comment by mysticmind** [comment]
@lamelyan If you could share a sample/code snippet of what you are doing especially in terms of using document store and document session in your code, it would help us to see what the issue could be. It is most likely how you are using it in your code paths.

**Reply by leo-amelyan** [reply]
@mysticmind  Thank you for your prompt response. I was able to track the problem down to a `using` statement. In a weird way, it's the `using` statement that somehow prevents the connection to be disposed of.  At least that's what I observed. The following is my code. 

In Startup's ConfigureContainer method, I call:

`services.IncludeRegistry<MartenServices>();`



The Marten service registry class:

```cs
public class MartenServices : ServiceRegistry
  {
      public MartenServices()
      {
          ForSingletonOf<IDocumentStore>().Use(c =>
          {
              return DocumentStore.For(options =>
              {
                  options.Connection(Configuration["Marten"]);
                  options.Events.StreamIdentity = StreamIdentity.AsString;
                  options.AutoCreateSchemaObjects = AutoCreate.All;
                  options.Tenancy.Cleaner.CompletelyRemoveAllAsync();
              });
          });

          // Register IDocumentSession as Scoped
          For<IDocumentSession>()
              .Use(c => c.GetInstance<IDocumentStore>().LightweightSession())
              .Scoped();

          // Register IQuerySession as Scoped
          For<IQuerySession>()
              .Use(c => c.GetInstance<IDocumentStore>().QuerySession())
              .Scoped();
      }
  }
```


The repository that executes the query.  

```cs
public class MyRepository : IMyRepository
  {
      readonly IUserInfo _userInfo;
      readonly IDocumentSession _documentSession;

      public MyRepository(IUserInfo userInfo, IDocumentSession documentSession)
      {
          _userInfo = userInfo;
          _documentSession = documentSession;
      }

      public async Task<IEnumerable<UserItem>> GetItemsForUser()
      {
          // The following assignment of the session variable is for demo purposes only...(yes, _documentSession can be used directly)
          // When the using statement is added, the db connections are not disposed of. 
          using var session = _documentSession;

         // When the using is removed, the number of connections does not go up. 
         // var session = _documentSession;

          return session.Query<UserItem>().Where(_ => _.UserId == _userInfo.Id);
      }
  }
```


Why does that happen?

**Reply by leo-amelyan** [reply]
Another question is about the DI configuration, [in the documentation](https://martendb.io/configuration/ioc.html), an instance of the `LightweightSession` is returned for IDocumentSession. 


```cs
 // Register IDocumentSession as Scoped
For<IDocumentSession>()
    .Use(c => c.GetInstance<IDocumentStore>().LightweightSession())
    .Scoped();
```

What if we want to use a different type of session (`OpenSession` or `DirtyTrackedSession`)?
